### PR TITLE
make jax.random default dtypes 64-bit

### LINKED ISSUE #756: Make samplers in jax.random use float64 as default when jax_enable_x64=True
Currently, the behavior of samplers in `jax.random` is somewhat inconsistent as compared to other array constructors when `jax_enable_x64=True` because it still uses `np.float32` as the default, unless the dtype argument is explicitly also set to `np.float64`. 

The following code demonstrates the issue:

```python
In [1]: from jax.config import config

In [2]: import jax.numpy as np

In [3]: from jax import random

In [4]: config.update('jax_enable_x64', True)

In [5]: np.array([1.]).dtype
/Users/npradhan/miniconda3/envs/numpyro/lib/python3.6/site-packages/jax/lib/xla_bridge.py:130: UserWarning: No GPU/TPU found, falling back to CPU.
  warnings.warn('No GPU/TPU found, falling back to CPU.')
Out[5]: dtype('float64')

In [6]: np.ones(1).dtype
Out[6]: dtype('float64')

In [7]: random.normal(random.PRNGKey(0), (1,)).dtype
Out[7]: dtype('float32')  # Expected: float64 (default as in other constructors)
```

This will be really useful to just change one line in the code and run all the computations in float64, instead of having to temporarily insert `dtype`s in calls to all the samplers. This should be a small change, and we are happy to put out a PR with the fix, but I was not sure if there are other design considerations involved. cc. @fehiepsi 

Related: #470.

### PR DESCRIPTION
**Pull request by mattjj**
fixes #756

### CODE REVIEW

File: jax/random.py [hunk 1]
```diff
@@ -412,14 +416,14 @@ def bernoulli(key, p=onp.float32(0.5), shape=()):
 def _bernoulli(key, p, shape):
   _check_shape("bernoulli", shape)
   shape = shape or onp.shape(p)
-  if not onp.issubdtype(onp.float32, lax.dtype(p)):
-    p = lax.convert_element_type(p, onp.float32)
+  if not onp.issubdtype(onp.float64, lax.dtype(p)):
+    p = lax.convert_element_type(p, onp.float64)
```
Comment (fehiepsi) [inline jax/random.py]: I think it is better to have `p=0.5` by default and do `p = lax.convert_element_type(p, dtype)` here with `dtype = xla_bridge.canonicalize_dtype(lax.dtype(p))`

File: jax/random.py [hunk 2]
```diff
@@ -371,17 +373,19 @@ def _shuffle(key, x, axis):
   return x
 
 
-def normal(key, shape, dtype=onp.float32):
+def normal(key, shape, dtype=onp.float64):
```
Comment (fehiepsi) [inline jax/random.py]: nit: similar to above, can we have a default `shape=()` here?

File: jax/random.py [hunk 3]
```diff
@@ -226,19 +226,21 @@ def _check_shape(name, shape):
     raise ValueError(msg.format(name, shape))
 
 
-def uniform(key, shape, dtype=onp.float32, minval=0., maxval=1.):
+def uniform(key, shape, dtype=onp.float64, minval=0., maxval=1.):
```
Comment (fehiepsi) [inline jax/random.py]: nit: can we have a default `shape=()` here

**Review by fehiepsi — COMMENTED** [review]
Thanks @mattjj ! All the changes look good to me. I just have a few minor comments on this. :)

File: jax/random.py (hunk 3 above)
Comment (mattjj) [reply jax/random.py]: Done!

File: jax/random.py (hunk 2 above)
Comment (mattjj) [reply jax/random.py]: Done!

File: jax/random.py (hunk 1 above)
Comment (mattjj) [reply jax/random.py]: I revised the logic a bit; PTAL and let me know what you think.

I think it should be an error if the `p` parameter is anything but a subdtype of `onp.floating`, so now we raise the error instead. I put the error-raising and casting outside the `@jit` subroutine for the weak reason that canonicalizing outside could get us more cache hits.

Comment (fehiepsi) [reply jax/random.py]: Yeah, the change looks great!

**Review by fehiepsi — APPROVED** [review]

File: jax/random.py (hunk 1 above)
Comment (mattjj) [reply jax/random.py]: Thanks for finding this, and checking it!

**Pull request merged**